proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+
+  * X does not implement equality.
+    Maybe you want to derive it using `deriving Eq`?
+  * In the expression: ...
+
+We definitely could include a generic framework to hint auto-derivation of built-in classes such as ``Eq`` or ``Ord``. But in the current Haskell ecosystem some libraries use generic deriving, others Template Haskell, and so on, so a generic facility seems useful.
+
+Inequalities
+~~~~~~~~~~~~
+The ability to attach a custom error message whenever an equality constraint fails is helpful in a variety of scenarios. The simpler one is simply explaining that two arguments should coincide. For example, if ``(==)`` is applied to terms of different type, we could say ::
+
+  * == is applied to arguments of different types `Bool` and `Int`
+  * In the expression: ...
+
+Another possibility is to give a description of the involved types in domain-specific terms. Take for example the `diagrams library <https://hackage.haskell.org/package/diagrams-core-1.4/docs/Diagrams-Core.html#t:QDiagram>`_, it defines a ``QDiagram b v n m`` type indexed by back-end, vector space, number field and annotation types. If we try to combine two diagrams of different vector space, we get a message similar to ::
+
+  * Couldn't match expected type 'V2' with actual type 'V3'
+
+Instead, we could express this message in a more meaningful way ::
+
+  * Couldn't match vector space 'V2' with 'V3'
+
+
+Proposed Change Specification
+-----------------------------
+This proposal defines two new "constraint combinators" to express constraint behavior in the event of a type error.
+
+The first combinator is ``IfUndischarged``. In short, ``IfUndischarged c msg`` shows the message when the constraint ``c`` is left undischarged by the constraint solver. A simple example using ``Eq`` ::
+
+    (==) :: IfUndischarged (Eq a) (ShowType a :<>: Text " does not implement equality."
+                                  :$$: Text "Maybe you want to derive it using `deriving Eq`?")
+         => a -> a -> Bool
+
+If it turns out that ``(==)`` is applied to a type for which no ``Eq`` instance exists, the error message becomes ::
+
+  * X does not implement equality.
+    Maybe you want to derive it using `deriving Eq`?
+  * In the expression: ...
+
+The second combinator is ``IfApartUnsafe`` (the naming shall become clear later). In short, ``IfApartUnsafe a b no yes`` rewrites to ``no`` if ``a`` and ``b`` are apart *at that point in the solving process* and to ``yes`` otherwise. Note that this is *extremely unsafe*, since any knowledge about ``a`` and ``b`` is lost if the second branch is taken, and for that reason we should only expose the variant ::
```

Comment:
This "at that point" is very subtle, and I think IIUC that @AntC2 has misunderstood this note. Here's my understanding:

The solver works by considering constraints in an impossible-to-predict order. While `IfUndischarged` makes a difference only at the very end of the solving process, when GHC has given up, `IfApartUnsafe` is different. It will check the apartness of two types at the time that the solver considers the constraint, reducing it to some other constraint depending on the result of the apartness test. Because it's possible for two non-apart types to become apart later in the solving process, `IfApartUnsafe` might choose the "not apart" result when a different ordering of constraints would have led to the "apart" result. This unpredictable behavior is (in my understanding) what's unsafe about `IfApartUnsafe`. As I see it, `IfApartUnsafe` can never launch the rockets; it can just cause unpredictable type inference, subtly depending on, say, the ordering of definitions in a file.

If I'm right here, the proposal should clarify the use of "unsafe" here, as it's different than many other uses of that term.